File: mobile/stores/goalStore.ts
import { create } from 'zustand';
import type {
  DailyGoal,
  GoalForDateResponse,
  GoalProfileListItem,
  UpdateGoalsForDateRequest,
} from '@shared/types';
import * as api from '@/services/api';

interface GoalState {
  goalsByDate: Record<string, DailyGoal | null>;
  metaByDate: Record<string, GoalForDateResponse | null>;
  profiles: GoalProfileListItem[];
  isLoading: boolean;
  error: string | null;
  fetch: (date: string) => Promise<void>;
  refreshProfiles: () => Promise<void>;
  saveChange: (data: UpdateGoalsForDateRequest) => Promise<void>;
}

export const useGoalStore = create<GoalState>((set, get) => ({
  goalsByDate: {},
  metaByDate: {},
  profiles: [],
  isLoading: false,
  error: null,

  fetch: async (date: string) => {
    set({ isLoading: true, error: null });
    try {
      const res = await api.getGoalsForDate(date);
      set((state) => ({
        goalsByDate: { ...state.goalsByDate, [date]: res.goals },
        metaByDate: { ...state.metaByDate, [date]: res },
        isLoading: false,
      }));
    } catch (e) {
      set({
        isLoading: false,
        error: e instanceof Error ? e.message : 'Failed to load goals',
      });
    }
  },

  refreshProfiles: async () => {
    try {
      const res = await api.getGoalProfiles();
      set({ profiles: res.profiles });
    } catch {
      // swallow; profiles are non-critical
    }
  },

  saveChange: async (data: UpdateGoalsForDateRequest) => {
    set({ isLoading: true, error: null });
    try {
      const res = await api.changeGoals(data);
      const date = res.date;
      set((state) => ({
        goalsByDate: { ...state.goalsByDate, [date]: res.goals },
        metaByDate: { ...state.metaByDate, [date]: res },
        isLoading: false,
      }));
      // Also refresh profiles so the list stays current
      const { refreshProfiles } = get();
      void refreshProfiles();
    } catch (e) {
      set({
        isLoading: false,
        error: e instanceof Error ? e.message : 'Failed to save goals',
      });
    }
  },
}));

File: MacroTrackerSwift/MacroTrackerSwift/Stores/GoalStore.swift
import Foundation
import Observation

/// Caches goal data keyed by date string.
/// Direct port of mobile/stores/goalStore.ts.
@Observable @MainActor
final class GoalStore {
    static let shared = GoalStore()

    var goalsByDate: [String: DailyGoal?] = [:]
    var metaByDate:  [String: GoalForDateResponse?] = [:]
    var profiles:    [GoalProfileListItem] = []
    var isLoading:   Bool = false
    var error:       String? = nil

    private init() {}

    func fetch(date: String) async {
        isLoading = true
        error = nil
        do {
            let res = try await APIClient.shared.getGoalsForDate(date)
            goalsByDate[date] = res.goals
            metaByDate[date]  = res
            isLoading = false
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }

    func refreshProfiles() async {
        do {
            let res = try await APIClient.shared.getGoalProfiles()
            profiles = res.profiles
        } catch {
            // non-critical — swallow
        }
    }

    func saveChange(data: UpdateGoalsForDateRequest) async {
        isLoading = true
        error = nil
        do {
            let res = try await APIClient.shared.changeGoals(data)
            goalsByDate[res.date] = res.goals
            metaByDate[res.date]  = res
            isLoading = false
            await refreshProfiles()
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }
}
